File: src/main/java/org/slimefunguguproject/bump/implementation/items/weapons/SkyDevilSword.java
package org.slimefunguguproject.bump.implementation.items.weapons;

import io.github.thebusybiscuit.slimefun4.api.recipes.RecipeType;
import org.bukkit.entity.DragonFireball;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.bukkit.potion.PotionEffect;
import org.bukkit.potion.PotionEffectType;
import org.bukkit.scheduler.BukkitRunnable;
import org.slimefunguguproject.bump.implementation.Bump;
import org.slimefunguguproject.bump.implementation.BumpItems;

public class SkyDevilSword extends BumpSword {

    public SkyDevilSword() {
        super(5, BumpItems.SKY_DEVIL_SWORD, RecipeType.ENHANCED_CRAFTING_TABLE, new ItemStack[] {
            null, null, null,
            BumpItems.SKY_SWORD, BumpItems.UPDATE_POWER, BumpItems.DEVIL_SWORD,
            null, null, null
        });
    }

    @Override
    public void onItemUse(Player p, ItemStack itemStack) {
        Bump.getLocalization().sendActionbarMessage(p, "weapon.sky_devil_sword.activate");

        p.setGlowing(true);
        p.addPotionEffect(new PotionEffect(PotionEffectType.SPEED, 300, 3));
        p.addPotionEffect(new PotionEffect(PotionEffectType.DAMAGE_RESISTANCE, 300, 3));

        (new BukkitRunnable() {
            int count = 3;

            @Override
            public void run() {
                if (count > 0) {
                    p.launchProjectile(DragonFireball.class);
                    count--;
                } else {
                    this.cancel();
                    p.setGlowing(false);
                    Bump.getLocalization().sendActionbarMessage(p, "weapon.sky_devil_sword.end");
                }
            }
        }).runTaskTimer(Bump.getInstance(), 1L, 100L);
    }
}

File: src/main/kotlin/io/github/slimefunguguproject/bump/implementation/items/weapons/SkyDevilSword.kt
package io.github.slimefunguguproject.bump.implementation.items.weapons

import io.github.slimefunguguproject.bump.Bump
import io.github.slimefunguguproject.bump.implementation.tasks.WeaponProjectileTask
import io.github.thebusybiscuit.slimefun4.api.items.ItemGroup
import io.github.thebusybiscuit.slimefun4.api.items.SlimefunItemStack
import io.github.thebusybiscuit.slimefun4.api.recipes.RecipeType
import org.bukkit.Sound
import org.bukkit.entity.DragonFireball
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack
import org.bukkit.potion.PotionEffect
import org.bukkit.potion.PotionEffectType
import org.bukkit.scheduler.BukkitRunnable

/**
 * [Heaven-demon Crumble Sword][SkyDevilSword] will fire 3 [dragon fireballs][DragonFireball]
 * and get some potion effects when using
 *
 * @author ybw0014
 */
class SkyDevilSword(
    itemGroup: ItemGroup,
    itemStack: SlimefunItemStack,
    recipeType: RecipeType,
    recipe: Array<ItemStack?>,
    hunger: Int,
) : BumpSword(itemGroup, itemStack, recipeType, recipe, hunger) {
    override fun onItemUse(p: Player, sword: ItemStack) {
        Bump.localization.sendActionbarMessage(p, "weapon.sky_devil_sword.activated")

        p.isGlowing = true
        p.addPotionEffect(PotionEffect(PotionEffectType.SPEED, 300, 3))
        p.addPotionEffect(PotionEffect(PotionEffectType.DAMAGE_RESISTANCE, 300, 3))

        object : BukkitRunnable() {
            var count: Int = 3

            override fun run() {
                if (count > 0) {
                    p.playSound(p.location, Sound.ENTITY_ENDER_DRAGON_SHOOT, 1.0f, 1.0f)
                    val projectile = p.launchProjectile(DragonFireball::class.java)
                    WeaponProjectileTask.track(projectile)
                    count--
                } else {
                    cancel()
                    p.isGlowing = false
                    Bump.localization.sendActionbarMessage(p, "weapon.sky_devil_sword.deactivated")
                }
            }
        }.runTaskTimer(Bump.instance, 1L, 100L)
    }
}
